ansible_base/models/resource.py

```diff
@@ -0,0 +1,132 @@
+# circular imports.
+def get_registry():
+    global _resource_registry
+    if _resource_registry is None:
+        from django.conf import settings
+
+        if hasattr(settings, "RESOURCE_REGISTRY_CONFIG"):
+            from django.utils.module_loading import import_string
+
+            _resource_registry = import_string(settings.RESOURCE_REGISTRY_CONFIG)
+        else:
+            _resource_registry = False
+    return _resource_registry
+
+
+class ResourceManager(models.Manager):
+    """
+    Ansible IDs are made up of two parts: the service short ID and a resource
+    UUIDv4. For efficiency reasons, these are stored as two separate fields:
+        - _service_id: Service that the resource belongs to.
+        - resource_id: Resource UUID.
+
+    Since Ansible ID is the value that will be referenced in the API, this
+    manager automatically annotates ansible_id as <service_short_id>:<resource_id>
+    for easier lookups.
+    """
+
+    def get_queryset(self):
+        # this needs to wait for migrations to run before it becomes available.
+        try:
+            s_id = service_id().split('-')[0]
+        except ProgrammingError:
+            s_id = "00000000"
+
+        return (
+            super()
+            .get_queryset()
+            .annotate(
+                _computed_service_id=models.Case(
+                    models.When(_service_id__isnull=False, then=models.F('_service_id')), default=models.Value(s_id), output_field=models.CharField()
+                ),
+                _ansible_id=Concat("_computed_service_id", models.Value(":"), models.F("resource_id"), output_field=models.CharField()),
+            )
+        )
+
+
+class ResourceType(models.Model):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.resource_registry = get_registry()
+
+    content_type = models.OneToOneField(ContentType, on_delete=models.CASCADE, related_name="resource_type", unique=True)
+    externally_managed = models.BooleanField()
+    migrated = models.BooleanField(null=False, default=False)
+    resource_type = models.CharField(max_length=256, unique=True, db_index=True)
+
+    @property
+    def serializer_class(self):
+        return self.get_resource_config()["managed_serializer"]
+
+    def get_resource_config(self):
+        return self.resource_registry.get_config_for_model(model=self.content_type.model_class())
+
+
+class Resource(models.Model):
+    objects = ResourceManager()
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, related_name="resources")
+
+    # this has to accommodate integer and UUID object IDs
+    object_id = models.TextField(null=False)
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    _service_id = models.CharField(null=True, max_length=8)
+
+    # we're not using this as the primary key because the ansible_id can change if the object is
+    # externally managed.
```

Comment:
`ansible_id can change if the object is externally managed.` <-- Can you mention some cases where the ansible_id can change? My mental model is that the id would not change.

Reply:
This can happen when merging resources across several services. For example if two services have the same group/team name, we are going to offer the admin to merge them into a single team. In this case, if the local instance already has an ansible_id for the resource, the ID will have to change to match the Gateway's ID.

I would kind of like to make the resource API available upstream, so that it can be used for RBAC, which is why an ansible ID might already exist.

With that said, I have been thinking about this problem this morning. It would be kind of nice to have the ansible ID be the PK for this table. It would allow us to use a one to one field on the resource model, and would let us serialize the ansible ID for resources without having to join on the Resource table or make a separate DB query, sp I'm open to the idea of changing this.

Reply:
I would try to avoid handing the `ansible_id can change` scenario ever, because I don't believe there's a business need for it. Yes, I think some teams will need to be merged. If the user doesn't want to merge them... then maybe we can offer a flag to control this in the migration. Before a service starts its own migration, couldn't it get a list of teams & ansible_ids from other services? I don't think that's technically problematic. Then we only need to define precedence of which services migrates first (or does this part of the migration) and passes on data to the others.